CMakeLists.txt

```diff
@@ -248,6 +252,50 @@ else(CMAKE_C_SIZEOF_DATA_PTR EQUAL 8)
   set(X64 OFF)
 endif (CMAKE_C_SIZEOF_DATA_PTR EQUAL 8)
 
+# The execution architecture, which might differ from the target for building
+# an AArch64 decoder to execute on x86 machines (i#1684).
+if (NOT "${TARGET_ARCH}" STREQUAL "${CMAKE_SYSTEM_PROCESSOR}")
+  set(DR_HOST_NOT_TARGET 1)
+  if (WIN32)
+    # TODO i#1648: Fix Windows compiler warnings for AArch64 on x86.
+    message(FATAL_ERROR "Host != target is not yet supported on Windows")
+  endif ()
+endif ()
+if (CMAKE_SYSTEM_PROCESSOR MATCHES "^arm")
+  set(DR_HOST_ARM 1)
+  set(DR_HOST_ARCH_NAME "arm")
+  set(DR_HOST_AARCHXX 1)
+  set(DR_HOST_ARCH_NAME_SHARED aarchxx)
+elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "^aarch64")
+  set(DR_HOST_AARCH64 1)
+  set(DR_HOST_ARCH_NAME "aarch64")
+  set(DR_HOST_AARCHXX 1)
+  set(DR_HOST_ARCH_NAME_SHARED aarchxx)
+  set(DR_HOST_X64 1)
+elseif (X64)
+  set(DR_HOST_X86 1)
+  set(DR_HOST_ARCH_NAME "x86")
+  set(DR_HOST_ARCH_NAME_SHARED ${DR_HOST_ARCH_NAME})
+  set(DR_HOST_X64 1)
+else ()
+  set(DR_HOST_X86 1)
+  set(DR_HOST_ARCH_NAME "x86")
+  set(DR_HOST_ARCH_NAME_SHARED ${DR_HOST_ARCH_NAME})
+endif ()
+
+if (DR_HOST_NOT_TARGET)
+  if (ARM)
+    set(X64 OFF)
+  else ()
+    set(X64 ON)
+  endif ()
+  if ((DR_HOST_X64 AND NOT X64) OR (NOT DR_HOST_X64 AND X64))
+    # XXX i#1345: We need to resolve the hardcoded ELF bitwidth types.
```

Comment:
There is a xref the other way but I put one in #1345 as well.  Note that I'm afraid it may still not work esp for a 32-bit host just with compiler type mismatches all over the place.